src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -1,58 +1,126 @@
+        {
+            get
+            {
+#if COMPLETION_LOCATION_TRACKING
+                return _completionLocation;
+#else
+                return null;
+#endif
+            }
+        }
+
+        public bool IsCompleted => Exception != null;
+
+        public PipeCompletionCallbacks TryComplete(Exception exception = null)
         {
 #if COMPLETION_LOCATION_TRACKING
             _completionLocation = Environment.StackTrace;
 #endif
-            // Set the exception object to the exception passed in or a sentinel value
-            Interlocked.CompareExchange(ref _exception, exception ?? _completedNoException, null);
+            if (Exception == null)
+            {
+                // Set the exception object to the exception passed in or a sentinel value
+                Exception = exception ?? _completedNoException;
+            }
+            return GetCallbacks();
         }
 
-        public bool IsCompletedOrThrow()
+        public PipeCompletionCallbacks AddCallback(Action<Exception, object> callback, object state)
         {
-            if (_exception != null)
+            if (_callbacks == null)
+            {
+                _callbacks = CompletionCallbackPool.Rent(InitialCallbacksSize);
+            }
+
+            var newIndex = _callbackCount;
+            _callbackCount++;
+
+            if (newIndex == _callbacks.Length)
+            {
+                var newArray = new PipeCompletionCallback[_callbacks.Length * 2];
+                Array.Copy(_callbacks, newArray, _callbacks.Length);
+                CompletionCallbackPool.Return(_callbacks);
+                _callbacks = newArray;
+            }
+            _callbacks[newIndex].Callback = callback;
+            _callbacks[newIndex].State = state;
+
+            if (IsCompleted)
             {
-                if (_exception != _completedNoException)
-                {
-                    ThrowFailed();
-                }
-                return true;
+                return GetCallbacks();
             }
-            return false;
+
+            return null;
         }
 
-        private void ThrowFailed()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsCompletedOrThrow()
         {
-            throw _exception;
+            if (Exception == null)
+            {
+                return false;
+            }
+
+            if (Exception != _completedNoException)
+            {
+                ThrowFailed();
+            }
+
+            return true;
         }
 
-        public string Location
+        private PipeCompletionCallbacks GetCallbacks()
         {
-            get
+            Debug.Assert(IsCompleted);
+            if (_callbackCount == 0)
             {
-#if COMPLETION_LOCATION_TRACKING
-                return _completionLocation;
-#else
                 return null;
-#endif
             }
+
+            var callbacks = new PipeCompletionCallbacks(CompletionCallbackPool, _callbackCount, Exception, _callbacks);
```

Comment:
You need to pass null for the exception if the exception is `_completedNoException`, right now this shows up as always having an exception.